### PR TITLE
Add support for the Phantom role

### PR DESCRIPTION
**Pull request by miniduikboot**
### Description

2024.6.18 adds the Phantom role, which is an Impostor that can vanish and appear. These mechanics use 4 new RPC's. This PR adds support for them and includes some basic anticheat checks

### CODE REVIEW

File: src/Impostor.Server/Net/Inner/Objects/InnerPlayerControl.cs [hunk 1]
```diff
@@ -1001,5 +1051,71 @@ private async ValueTask<bool> HandleSetStartCounter(ClientPlayer sender, int seq
 
             return true;
         }
+
+        private async ValueTask<bool> HandleCheckVanish(ClientPlayer sender, float maxDuration)
+        {
+            if (!await ValidateRole(RpcCalls.CheckVanish, sender, PlayerInfo, RoleTypes.Phantom))
```
Comment (NikoCat233) [inline src/Impostor.Server/Net/Inner/Objects/InnerPlayerControl.cs:1057]: ValidateRole got used twice in these checks.

Comment (NikoCat233) [inline src/Impostor.Server/Net/Inner/Objects/InnerPlayerControl.cs:1057]: ValidateRole got used twice in these checks

**Review by NikoCat233 — CHANGES_REQUESTED** [review]

File: src/Impostor.Server/Net/Inner/Objects/InnerPlayerControl.cs (hunk 1 above)
Comment (miniduikboot) [reply src/Impostor.Server/Net/Inner/Objects/InnerPlayerControl.cs]: Good point, fixed

**Review by NikoCat233 — APPROVED** [review]

File: src/Impostor.Api/Net/Messages/Rpcs/Rpc62CheckVanish.cs [hunk 2]
```diff
@@ -0,0 +1,15 @@
+namespace Impostor.Api.Net.Messages.Rpcs
+{
+    public static class Rpc62CheckVanish
+    {
+        public static void Serialize(IMessageWriter writer, float maxDuration = 0f)
```
[the remark this answers is not in the document]
Comment (Pietrodjaowjao) [reply src/Impostor.Api/Net/Messages/Rpcs/Rpc62CheckVanish.cs:5]: Yeah, remove default

File: src/Impostor.Server/Net/Inner/Objects/InnerPlayerControl.cs [hunk 3]
```diff
@@ -1008,5 +1058,61 @@ private async ValueTask<bool> HandleSetStartCounter(ClientPlayer sender, int seq
 
             return true;
         }
+
+        private async ValueTask<bool> HandleCheckVanish(ClientPlayer sender, float maxDuration)
+        {
+            // TODO: Check if operation is taking place during lobby/meetings
```
[the remark this answers is not in the document]
Comment (Pietrodjaowjao) [reply src/Impostor.Server/Net/Inner/Objects/InnerPlayerControl.cs:1064]: HandleCheckVanishAsync

[the remark this answers is not in the document]
Comment (miniduikboot) [reply src/Impostor.Server/Net/Inner/Objects/InnerPlayerControl.cs:1064]: 6pak: I considered TODO's are out of scope of this PR, as the intent was just to quickly bring up the Phantom role

Pietro: None of the other HandleX methods end their name on Async. Also please don't hijack review threads like this

File: src/Impostor.Server/Net/Inner/Objects/InnerPlayerControl.cs [hunk 4]
```diff
@@ -1008,5 +1058,61 @@ private async ValueTask<bool> HandleSetStartCounter(ClientPlayer sender, int seq
...
+            // TODO: Check max duration
```
[the remark this answers is not in the document]
Comment (miniduikboot) [reply src/Impostor.Server/Net/Inner/Objects/InnerPlayerControl.cs:1065]: the idea was that someone should add a check if the maximum duration of a vanish is not being exceeded

[the remark this answers is not in the document]
Comment (js6pak) [reply src/Impostor.Server/Net/Inner/Objects/InnerPlayerControl.cs:1065]: I thought this TODO was related to the `maxDuration` parameter, but if it's a timing check then I agree we can do that later.

**Review by js6pak — APPROVED** [review]

**Pull request merged**